### PR TITLE
perf: extend type inference to class methods and fix object-to-double local store

### LINKED ISSUE #129: Track variable type provenance for optimization opportunities
## Overview

Implement variable type provenance tracking throughout the compilation pipeline to enable future optimizations that reduce unnecessary boxing/unboxing operations.

## Problem Statement

Currently, js2il treats all JavaScript values as `System.Object` in the generated IL, leading to:

1. **Excessive Boxing**: Primitive values (numbers, booleans) are boxed on every operation
2. **Runtime Overhead**: Type checks and conversions happen at runtime via `TypeUtilities.ToNumber()`, etc.
3. **Missed Optimizations**: Cannot use native .NET arithmetic when types are statically known
4. **Poor Performance**: Numeric hot paths (loops, math operations) suffer from allocation pressure

Example current IL for `x + 1`:
```il
ldloc.0        // Load x as object
ldc.r8 1.0     // Load 1.0 as double
box [System.Runtime]System.Double
call object [JavaScriptRuntime]JavaScriptRuntime.Operators::Add(object, object)
// Returns boxed result
```

## Proposed Solution

Track the **provenance** (origin and type history) of each variable and expression throughout compilation to statically determine types when possible.

### Type Provenance Information

```csharp
public class TypeProvenance
{
    // Static type if known with certainty
    public JavascriptType? KnownType { get; set; }
    
    // Possible types if narrowed (e.g., "number | string")
    public HashSet<JavascriptType> PossibleTypes { get; set; }
    
    // Where the value originated
    public ProvenanceSource Source { get; set; }
    
    // Confidence level for optimizations
    public ProvenanceConfidence Confidence { get; set; }
}

public enum ProvenanceSource
{
    Literal,           // const x = 42
    Parameter,         // function f(x) - unknown
    BinaryOp,          // x + y - depends on operands
    FunctionReturn,    // const x = foo() - depends on foo's return type
    TypeCoercion,      // Number(x) - known to be number
    ArrayElement,      // arr[0] - unknown
    ObjectProperty     // obj.prop - unknown
}

public enum ProvenanceConfidence
{
    Certain,   // Literal 42 is definitely a number
    Likely,    // Result of x + y where both are numbers
    Possible,  // Parameter might be a number based on usage
    Unknown    // No information
}
```

### Phase 1: Literal Tracking (MVP)
Track types for literals and constants:

```javascript
const x = 42;        // KnownType = Number
const y = "hello";   // KnownType = String
const z = true;      // KnownType = Boolean
```

Benefits:
- Optimize arithmetic with literal operands
- Skip boxing for primitive literals
- Emit native IL operations when safe

### Phase 2: Expression Type Inference
Propagate types through expressions:

```javascript
const a = 10;        // Number
const b = 20;        // Number
const c = a + b;     // Number (both operands are Number)
const d = c * 2;     // Number (Number * Number)
```

Rules:
- `Number + Number` → `Number`
- `String + Any` → `String`
- `Number - Number` → `Number`
- etc.

### Phase 3: Control Flow Analysis
Track types through conditionals and loops:

```javascript
let x = 10;          // Number
if (condition) {
    x = 20;          // Still Number
} else {
    x = "hello";     // Now String - type widened to Number | String
}
// After if: x is Number | String
```

Need to:
- Compute type unions at merge points
- Track narrowing in branches
- Handle type guards (`typeof x === 'number'`)

### Phase 4: Function Return Type Inference
Infer function return types from body:

```javascript
function add(a, b) {
    return a + b;    // If called with numbers, returns number
}

const result = add(10, 20);  // result is Number
```

This requires:
- Call-site analysis
- Return statement analysis
- Multiple return path handling

### Phase 5: Inter-procedural Analysis
Track types across function boundaries:

```javascript
function double(x) {
    return x * 2;    // x must be coercible to number
}

const y = double(21); // y is Number
```

### 1. Symbol Table
Extend `Variable` class:

```csharp
public class Variable
{
    // ... existing members
    public TypeProvenance Provenance { get; set; } = new();
}
```

### 2. Type Inference Pass
Add new compilation phase after symbol table building:

```
Parse → Validate → SymbolTable → **TypeInference** → TypeGeneration → IL Emission
```

```csharp
public class TypeInferenceAnalyzer
{
    private readonly SymbolTable _symbolTable;
    
    public void AnalyzeProgram(Program program)
    {
        // Visit all nodes and assign provenance
        foreach (var statement in program.Body)
            AnalyzeStatement(statement);
    }
    
    private TypeProvenance AnalyzeExpression(Expression expr)
    {
        return expr switch
        {
            Literal lit => AnalyzeLiteral(lit),
            BinaryExpression bin => AnalyzeBinaryExpression(bin),
            CallExpression call => AnalyzeFunctionCall(call),
            // ... other cases
        };
    }
}
```

### 3. IL Generation
Use provenance to emit optimized IL:

```csharp
public void EmitBinaryOperation(BinaryExpression expr)
{
    var leftType = GetProvenance(expr.Left);
    var rightType = GetProvenance(expr.Right);
    
    if (leftType.KnownType == JavascriptType.Number && 
        rightType.KnownType == JavascriptType.Number)
    {
        // Emit native double arithmetic
        EmitExpression(expr.Left, unbox: true);
        EmitExpression(expr.Right, unbox: true);
        _il.OpCode(ILOpCode.Add);  // Native add, no boxing
        // Box only if result needs to be stored as object
    }
    else
    {
        // Fall back to runtime operators
        EmitExpression(expr.Left, box: true);
        EmitExpression(expr.Right, box: true);
        _runtime.InvokeOperatorsAdd();
    }
}
```

### 1. Unboxed Arithmetic
When both operands are known to be numbers:

```javascript
const x = 10;
const y = 20;
const z = x + y;
```

Emit:
```il
ldc.r8 10.0
ldc.r8 20.0
add
stloc.0  // Store as unboxed double
```

Instead of:
```il
ldc.r8 10.0
box [System.Runtime]System.Double
ldc.r8 20.0
box [System.Runtime]System.Double
call object Operators::Add(object, object)
```

### 2. Loop Optimization
Hot paths benefit most:

```javascript
for (let i = 0; i < 1000; i++) {  // i is Number
    sum = sum + i;                 // sum is Number
}
```

Use native int32 operations instead of boxed doubles.

### 3. Type-Specialized Variables
Store variables with known types unboxed:

```csharp
// Instead of all variables as object fields
public class Scope {
    public object x;
    public object y;
}

// Generate typed fields when possible
public class Scope {
    public double x;  // Known to be number
    public object y;  // Unknown type
}
```

### 4. Eliminated Type Checks
Skip runtime coercion when type is known:

```javascript
const x = 42;
const y = Number(x);  // Redundant - x is already a number
```

Can eliminate the `Number()` call.

### Correctness Tests
Ensure optimizations don't change semantics:

1. **Type Widening**: Variables with multiple types handled correctly
2. **Type Narrowing**: Conditional branches maintain type safety
3. **Edge Cases**: `undefined`, `null`, `NaN` handled properly
4. **Operator Semantics**: `+` with strings vs numbers

### Performance Tests
Measure impact:

1. **Microbenchmarks**: Arithmetic loops, array operations
2. **Prime Sieve**: Should see significant improvement
3. **Memory Pressure**: Reduced allocations

### Regression Tests
Ensure no correctness regressions on existing test suite.

## Metrics & Goals

Success criteria:
- **10-50% reduction** in allocations for numeric code
- **2-5x speedup** on math-heavy benchmarks (prime sieve)
- **Zero correctness regressions** on existing tests
- **Minimal compile time increase** (< 10%)

### 1. JavaScript Semantics
JavaScript's dynamic typing is subtle:
- `"5" - 2` → `3` (string coerced to number)
- `"5" + 2` → `"52"` (number coerced to string)
- `null + 1` → `1` (null → 0)

Must be conservative to maintain correctness.

### 2. Scope Complexity
Variables can have different types in different scopes:

```javascript
let x = 10;
function foo() {
    x = "hello";  // Same variable, different type
}
```

Need inter-procedural analysis or conservative fallback.

### 3. Complexity Budget
Type inference adds compilation complexity. Balance accuracy vs. compile time.

### 4. Maintenance
More sophisticated IL generation = more code to maintain and test.

## Future Enhancements

Beyond initial implementation:

1. **Type Guards**: Honor `typeof` checks for narrowing
2. **Generic Specialization**: Generate multiple versions of functions for different type arguments
3. **Profile-Guided Optimization**: Use runtime profiling to guide type decisions
4. **SIMD Opportunities**: Vectorize array operations when elements are known numbers
5. **Int32 Optimization**: Use int32 instead of double when appropriate

## Documentation

- Document type inference algorithm in `docs/TypeInference.md`
- Update architecture docs with new compilation phase
- Add examples of optimized vs unoptimized IL
- Performance tuning guide for users

## Related Work

- May benefit from #122 (GlobalThis instance) for clearer lifetime semantics
- Should track types in `VariableRegistry` from Type Generation phase
- Consider impact on future JIT compilation (#future)

## References

- [V8 Type Inference](https://v8.dev/blog/turbofan-jit)
- [TypeScript Type Inference](https://www.typescriptlang.org/docs/handbook/type-inference.html)
- [Flow Type Inference](https://flow.org/en/docs/lang/types-and-expressions/)
- [C# JIT Optimizations](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/ryujit-overview.md)
- [Optimization Techniques](https://dl.acm.org/doi/10.1145/255669.255671)

### LINKED ISSUE #141: feat: type inference for bitwise operators
## Description

Extend type inference to recognize that bitwise operators always return numbers in JavaScript. This will enable unboxed `float64` storage for variables assigned from bitwise operations, improving performance.

## Currently Failing Tests

Type inference tests in `SymbolTableTypeInferenceTests.cs`:
- `5 & 3` (AND)
- `5 | 3` (OR)
- `5 ^ 3` (XOR)
- `5 << 2` (Left shift)
- `5 >> 2` (Signed right shift)
- `5 >>> 2` (Unsigned right shift)
- `~5` (NOT - unary)

## Implementation Tasks

1. **Extend `InferExpressionClrType` method** in `Js2IL/SymbolTable/SymbolTableBuilder.InferTypes.cs`:
   - Handle `NonLogicalBinaryExpression` with bitwise operators (`BitwiseAnd`, `BitwiseOr`, `BitwiseXor`, `LeftShift`, `RightShift`, `UnsignedRightShift`) → return `typeof(double)`
   - Handle `UnaryExpression` with `BitwiseNot` operator → return `typeof(double)`

2. **Implement IL generation for Bitwise NOT** (`~` operator):
   - Currently throws "Unsupported unary operator: BitwiseNot"
   - Add support in `ILExpressionGenerator.EmitUnaryExpression`
   - Test case already exists: `UnaryOperator_BitwiseNot` (currently skipped)

3. **Update tests**:
   - All type inference tests should pass
   - `UnaryOperator_BitwiseNot` execution and generator tests should pass

## Expected Performance Impact

Variables like `const wordOffset = index >>> 5` in PrimeJavaScript benchmark will use unboxed `float64` locals instead of boxed `object` locals, eliminating boxing/unboxing overhead.

## References

- ECMAScript spec: Bitwise operators always convert operands to 32-bit integers and return Number type
- Related PR #140: Extended type inference to class methods

### PR DESCRIPTION
**Pull request by tomacox74**
## Summary

This PR extends the type inference optimization to class methods and their nested block scopes, and fixes several critical bugs related to typed local variables. It significantly improves performance for numeric-intensive code in classes.

Resolves #141 (type inference for bitwise operators)
Resolves #129 (variable type provenance tracking - phase 1: class methods)

#### Type Inference for Class Methods
- Extended `SymbolTableBuilder.InferTypes` to recursively process child scopes including class methods
- Uncaptured numeric variables in class methods now use unboxed `float64` locals instead of boxed objects
- Eliminates boxing/unboxing overhead in numeric computations

#### Type Inference for Block Scopes in Class Methods
- Type inference now applies to variables in for loops, while loops, if/else blocks, try/catch/finally, and switch cases within class methods
- Added `isBlockScopeInClassMethod()` helper that walks up the scope tree to verify the block is within a class method without crossing intermediate function boundaries
- Enables unboxed locals for loop iterators and intermediate calculations in nested control flow

#### Type Inference for Bitwise Operators
- Extended `InferExpressionClrType` to recognize bitwise operators (`&`, `|`, `^`, `<<`, `>>`, `>>>`, `~`) always return numbers
- Variables assigned from bitwise operations now use unboxed `float64` storage

#### Object-to-Double Local Store
- Fixed IL generation when assigning method call results to stable typed double locals
- Added `TypeUtilities.ToNumber()` call to convert object references to double before storing
- Fixed `valueIsBoxed` tracking to correctly distinguish unboxed float64 values from boxed object values

#### Block-Scope Local Variable Type Lookup
- Fixed `GetLocalVariableType` failing to find block-scope uncaptured variables
- Added `_blockScopeLocalsBySlot` dictionary cache, populated by `TryResolveFieldBackedVariable`
- Block-scope variables now correctly return their inferred CLR type instead of defaulting to `System.Object`
- **Impact**: `setBitsTrue` (PrimeJavaScript benchmark) now has 9 `float64` locals instead of 1

#### Compound Assignment in Class Methods
- Fixed incorrect IL generation for compound assignments (`+=`, `|=`) in class instance methods
- Instance methods now correctly load parent scopes from the `scopes` parameter array using `ldarg` + `ldelem_ref` + `castclass`

#### New Unit Tests for Variables Class (12 tests)
- Block-scope variable resolution with correct CLR type propagation
- Unique local slot allocation for variables in different block scopes
- Nested block scope shadowing
- `GetLocalVariableType` correctly returning `float64` for stable double-typed block-scope locals
- Integration test simulating `setBitsTrue` pattern

#### Block-Scope Type Inference Tests (15 tests)
- For loops, while loops, do-while loops, for...in, for...of
- If/else blocks, try/catch/finally blocks, switch cases
- Nested block scopes (2-3 levels deep)

#### Other
- Added `InternalsVisibleTo` to expose internal classes to test project
- Added `Classes_ClassMethod_LocalVar_ReassignedFromMethodCall` test for object-to-double conversion

## Before/After Example

**JavaScript:**
```javascript
class C { run() { let i = 0; const q = 10; while (i < q) { i = i + 1; } return i; } }
```

**Before (boxed):**
```il
.locals init ([0] object, [1] object)
IL_0000: ldc.r8 0.0
IL_0009: box System.Double
IL_000e: stloc.0
```

**After (unboxed):**
```il
.locals init ([0] float64, [1] float64)  
IL_0000: ldc.r8 0.0
IL_0009: stloc.0
```

## Performance Impact

- Numeric loops in class methods now use native double arithmetic
- Eliminates boxing allocation per operation
- `setBitsTrue` method: 9 unboxed `float64` locals (was 1)
- Significant improvement for bitwise operations in nested loops

## Related Issues
- Created #142 (extend type inference to functions and arrow functions)
- Created #143 (add bool and string typed local variable support)

### CODE REVIEW

File: Js2IL/SymbolTable/SymbolTableBuilder.InferTypes.cs [hunk 1]
```diff
@@ -9,6 +9,47 @@ public partial class SymbolTableBuilder
     /// </Summary>
     private void InferVariableClrTypes(Scope scope)
     {
+        InferVariableClrTypesRecursively(scope);
+    }
+
+    void InferVariableClrTypesForScope(Scope scope)
+    {
+        // these are temporary gates as we slowing roll out type inference
```
[the remark this answers is not in the document]
Comment (tomacox74) [reply Js2IL/SymbolTable/SymbolTableBuilder.InferTypes.cs]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/tomacox74/js2il/pull/140#discussion_r2621105263)
